requestmatch/matcher.go

```diff
@@ -0,0 +1,362 @@
+	"sort"
+)
+
+const freeWildcardExp = "/[*][^/]+$"
+
+type leafMatcher struct {
+	method         string
+	hostRxs        []*regexp.Regexp
+	pathRxs        []*regexp.Regexp
+	headersExact   map[string]string
+	headersRegexps map[string][]*regexp.Regexp
+	value          interface{}
+}
+
+type leafMatchers []*leafMatcher
+
+type pathMatcher struct {
+	leaves            leafMatchers
+	freeWildcardParam string
+}
+
+// A Matcher represents a preprocessed set of definitions and their associated
+// values.
+type Matcher struct {
+	paths               *pathmux.Tree
+	rootLeaves          leafMatchers
+	ignoreTrailingSlash bool
+}
+
+// A Definition represents a set of conditions and an associated
+// value to be returned when a request fulfills all non-empty
+// conditions.
+type Definition interface {
+
+	// Fixed or wildcard path, or empty.
+	Path() string
+
+	// Method to match, or empty.
+	Method() string
+
+	// Regular expressions, matched if all matched by the `Host` field.
+	HostRegexps() []string
+
+	// Regular expressions, matched if all matched by the request path.
+	PathRegexps() []string
+
+	// Exact matches for request headers.
+	Headers() map[string]string
+
+	// Regular expressions for header fields, matched if all matched.
+	HeaderRegexps() map[string][]string
+
+	// Associated value returned in case of a match.
+	Value() interface{}
+}
+
+// An error created if a definition cannot be preprocessed.
+type DefinitionError struct {
+	Index    int
+	Original error
+}
+
+var freeWildcardRx *regexp.Regexp
+
+func init() {
+	freeWildcardRx = regexp.MustCompile(freeWildcardExp)
+}
+
+func compileRxs(exps []string) ([]*regexp.Regexp, error) {
+	rxs := make([]*regexp.Regexp, len(exps))
+	for i, exp := range exps {
+		rx, err := regexp.Compile(exp)
+		if err != nil {
+			return nil, err
+		}
+
+		rxs[i] = rx
+	}
+
+	return rxs, nil
+}
+
+func makeLeaf(d Definition) (*leafMatcher, error) {
+	hostRxs, err := compileRxs(d.HostRegexps())
+	if err != nil {
+		return nil, err
+	}
+
+	pathRxs, err := compileRxs(d.PathRegexps())
+	if err != nil {
+		return nil, err
+	}
+
+	headerExps := d.HeaderRegexps()
+	allHeaderRxs := make(map[string][]*regexp.Regexp)
+	for k, exps := range headerExps {
+		headerRxs, err := compileRxs(exps)
+		if err != nil {
+			return nil, err
+		}
+
+		allHeaderRxs[k] = headerRxs
+	}
+
+	return &leafMatcher{
+		method:         d.Method(),
+		hostRxs:        hostRxs,
+		pathRxs:        pathRxs,
+		headersExact:   d.Headers(),
+		headersRegexps: allHeaderRxs,
+		value:          d.Value()}, nil
+}
+
+func freeWildcardParam(path string) string {
+	param := freeWildcardRx.FindString(path)
+	if param == "" {
+		return ""
+	}
+
+	// clip '/*' and return only the name
+	return param[2:]
+}
+
+// Constructs a Matcher based on the provided definitions. If `ignoreTrailingSlash`
+// is true, the Matcher handles paths with or without a trailing slash equally.
+func Make(ds []Definition, ignoreTrailingSlash bool) (*Matcher, []*DefinitionError) {
```

Comment:
you're right. I also thought of it. Since now this branch is about the mailgun replacement, I think we can leave it like this. I already have a refactoring branch, and I will add this to the todos for that.

As the idiomatism goes and how I noticed recently, Make* is rarelier used and is usually applied to struct values only, recalling the builtin make, while New* is more often used, and I find it now more appropriate to pointer values, too. This aligns with the fact, that we have an almost never used builtin `new`, where new(Type) is equivalent to &Type{}.

Anyway, let's let this in like this, and then I'll fix this as a refactoring effort.